sample-app/src/service-worker.ts

```diff
@@ -0,0 +1,71 @@
+/// <reference lib="webworker" />
+import { clientsClaim } from 'workbox-core';
+import { precacheAndRoute } from 'workbox-precaching';
+
+declare const self: ServiceWorkerGlobalScope;
+
+clientsClaim();
+const precacheManifest = self.__WB_MANIFEST;
+// Only precache static assets in production.
+// Precaching in development can result in confusing situations.
+// See https://github.com/facebook/create-react-app/issues/2398#issuecomment-304638935
+if (process.env.NODE_ENV !== 'development') {
+  precacheAndRoute(precacheManifest);
```

Comment:
Out of curiosity, how much does pre-caching the manifest get us? Could it make the installation of the service worker more brittle than if we didn't cache any static assets?

Reply:
Precaching static assets speeds up subsequent visits, and also allows for offline support. We get nearly instant load speeds for static js/css, which is kind of scope creep on this pr but seemed way too juicy to pass up. (I'm very excited about this)

![image](https://user-images.githubusercontent.com/23005393/145427932-eaa85ad1-20d7-4cca-9f0b-ce0ba481359e.png)

Precaching doesn't affect the brittleness of the service worker, but it does introduce a little complexity on making sure clients always have the most up to date assets. I'm glad you asked me about this. I did some testing, and out of the box, updates to the service worker will only be activated once all pages using the old service worker have been closed, and the page would also have to be refreshed an additional time to use the updated precached assets. This aligns with what all of the service worker documentation I read on the web says.

To remedy this I added a `self.skipWaiting()` to the install step as well as an autorefresh to the service worker registration. This is based on what the create-react-app PWA template does. The autorefresh seems like a hack, but because of the precached assets the refresh occurs almost instantly.